sound/soc/sof/intel/hda.c

```diff
@@ -1039,13 +1039,72 @@ static int hda_generic_machine_select(struct snd_sof_dev *sdev)
 #endif
 
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL_SOUNDWIRE)
+/* Check if all slaves defined on the link can be found */
+static bool link_slaves_found(struct snd_sof_dev *sdev,
+			      struct snd_soc_acpi_link_adr *link,
+			      struct sdw_intel_ctx *sdw)
+{
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	struct sdw_intel_slave_id *ids = sdw->ids;
+	int num_slaves = sdw->num_slaves;
+	unsigned int part_id, link_id, unique_id, mfg_id;
+	int j, k;
+
+	/*
+	 * Spec definition
+	 *   Register           Bit     Contents
+	 *   DevId_0 [7:4]      47:44   sdw_version
+	 *   DevId_0 [3:0]      43:40   unique_id
+	 *   DevId_1            39:32   mfg_id [15:8]
+	 *   DevId_2            31:24   mfg_id [7:0]
+	 *   DevId_3            23:16   part_id [15:8]
+	 *   DevId_4            15:08   part_id [7:0]
+	 *   DevId_5            07:00   class_id
+	 */
+	for (k = 0; k < link->num_adr; k++) {
+		mfg_id = (link->adr[k] >> 24) & GENMASK(15, 0);
+		part_id = (link->adr[k] >> 8) & GENMASK(15, 0);
+		for (j = 0; j < num_slaves; j++) {
```

Comment:
@bardliao the logic for these loops is correct but not very intuitive. You are trying to find if all the slaves in sdw->ids are present on the link, right? Then the first loop should be for(i=0; i<num_slaves;i++) followed by the for(j=0;  j<link->num_adr;j++). But I guess its fine both ways logically.

Reply:
@ranj063 I am trying to check the links one by one and check all slaves on the link. That's why I check link outside and slave inside.